src/librustc/traits/select.rs

```diff
@@ -1522,6 +1522,33 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
             .map(|v| v.get(tcx))
     }
 
+    /// Determines whether can we safely cache the result
+    /// of selecting an obligation. This is almost always 'true',
+    /// except when dealing with certain ParamCandidates.
+    ///
+    /// Ordinarily, a ParamCandidate will contain no inference variables,
+    /// since it was usually produced directly from a DefId. However,
+    /// certain cases (currently only librustdoc's blanket impl finder),
+    /// a ParamEnv may be explicitly constructed with inference types.
+    /// When this is the case, we do *not* want to cache the resulting selection
+    /// candidate. This is due to the fact that it might not always be possible
+    /// to equate the obligation's trait ref and the candidate's trait ref,
+    /// if more constraints end up getting added to an inference variable.
+    ///
+    /// Because of this, we always want to re-run the full selection
+    /// process for our obligation the next time we see it, since
+    /// we might end up picking a different SelectionCandidate (or none at all)
+    fn can_cache_candidate(&self,
+        result: &SelectionResult<'tcx, SelectionCandidate<'tcx>>
+     ) -> bool {
+        match result {
+            Ok(Some(SelectionCandidate::ParamCandidate(trait_ref))) => {
+                !trait_ref.skip_binder().input_types().any(|t| t.walk().any(|t_| t_.is_ty_infer()))
```

Comment:
This is equivalent to `trait_ref.needs_infer()`, but potentially much slower (since methods like `needs_infer` rely on precomputed flags), I wonder how much of a perf regression this is.

Both `TraitRef::input_types` and `Ty::walk` are warning flags, we might want to audit their uses (and make `Ty::walk` more annoying to use, probably).

cc @nikomatsakis @pnkfelix @matthewjasper

Reply:
Yes, @eddyb is correct, this does seem slower (I hadn't noticed this PR at the time, though I agree also with @arielb1's comment that it's a bit surprising to see inference variables in the param env, though I suppose there may be some path by which it can occur -- and eventually it'll be true)

Reply:
(I don't think I agree that `input_types` is a warning flag, though -- `walk` maybe is..?)

Reply:
Anyway, should we open an issue to try converting this to use the `needs_infer` call?

Reply:
It's included in https://github.com/rust-lang/rust/pull/69294#discussion_r383403240, I haven't had the chance to split that into its own PR for perf testing (presumably we can just land it with `rollup=never` and see the results there. it can't be slower than `walk`-ing, anyway).

Reply:
> I don't think I agree that `input_types` is a warning flag

Isn't it a leftover from the days of "output" modes on parameters? Like before associated types fully became a thing?

Also it's going to get really bad with `const` generics, so @varkor or @yodaldevoid would probably have to remove `input_types` just to make its callsites correct.

Reply:
> Isn't it a leftover from the days of "output" modes on parameters? Like before associated types fully became a thing?

It seems like it just screens out lifetime parameters. I guess you're right that there is probably generally little reason to do this.

Reply:
> It seems like it just screens out lifetime parameters

And const parameters :fearful: (presumably), hence my comments on that.

Reply:
I see. And I suppose it's probably just wrong to ignore needs-infer in lifetime parameters anyway. OK, I agree it's probably a 'red flag' of some kind.

Reply:
But it seems like you're going to fix this in #69294, so that's good.